### PR TITLE
Bar Gauge: Add field name placement option

### LINKED ISSUE #75928: Bar Gauge: Add option for label placement
### Discussed in https://github.com/grafana/grafana/discussions/37026

<div type='discussions-op-text'>

<sup>Originally posted by **mxmorin** February 17, 2021</sup>
**What happened**:
When my query returns several lines, the legend is placed in front of the bars.
When it returns only one line (display name is valued at ${__series.name}), the legend is placed above the bars.

image

Is it possible to have the same placement whatever the return and better, provide a field to set it up ?

**What you expected to happen**:
Legend is place at the same place regardless of the number of lines returned or have a field to set the placement

**Environment**:
- Grafana version: 7.4.1
- Data source type & version: Prometheus
- OS Grafana is installed on: Centos 7
- User OS & Browser: Chrome
</div>

### PR DESCRIPTION
**Pull request by nmarrs**
Updated behavior (with `namePlacement`)

https://github.com/grafana/grafana/assets/22381771/5896c4d2-4915-4383-96c7-90196e17aaf8

Auto (before)

https://github.com/grafana/grafana/assets/22381771/cdb87f30-b725-44bb-b708-f38710964294

Auto / Top / Left

https://github.com/grafana/grafana/assets/22381771/f2d446a1-da7a-4eaa-b8e5-179b25673c39

Fixes https://github.com/grafana/grafana/issues/75928

### CODE REVIEW

File: docs/sources/panels-visualizations/visualizations/bar-gauge/index.md [hunk 1]
```diff
@@ -67,6 +67,22 @@ Choose a display mode.
...
 
+### Value display
+
+Choose a value display mode.
+
+- **Value color -** Value color is determined by value.
+- **Text color -** Value color is default text color.
+- **Hidden -** Values are hidden.
+
+### Value placement
```
Comment (mckn) [inline docs/sources/panels-visualizations/visualizations/bar-gauge/index.md:78]: Question: I might have misunderstood this feature but I feel that the naming of this option is a bit confusing. I thought it was the field name that you configured the placement for and not the value?

**Review by mckn — COMMENTED** [review]
Code looks great! Added a question about the naming.

**Comment by torkelo** [comment]
@nmarrs should not be named "name placement" option? it does not look like it's changing the value placement

File: docs/sources/panels-visualizations/visualizations/bar-gauge/index.md [hunk 2]
```diff
@@ -67,6 +67,22 @@ Choose a display mode.
...
+
+Choose a value placement mode. Note: This option only applies when the orientation of the bar gauge is horizontal. When the bar gauge is in the vertical orientation, values are always placed at the bottom of each bar gauge.
```
Comment (imatwawana) [inline docs/sources/panels-visualizations/visualizations/bar-gauge/index.md:80]:
```suggestion
Choose a value placement mode. 

{{% admonition type="note" %}}
This option only applies when the orientation of the bar gauge is horizontal. When the bar gauge is in the vertical orientation, values are always placed at the bottom of each bar gauge.
{{% /admonition %}}
```

File: docs/sources/panels-visualizations/visualizations/bar-gauge/index.md [hunk 3]
```diff
@@ -67,6 +67,22 @@ Choose a display mode.
...
+
+- **Auto -** Grafana selects what it thinks is the best placement.
```
Comment (imatwawana) [inline docs/sources/panels-visualizations/visualizations/bar-gauge/index.md:82]:
```suggestion
- **Auto -** Grafana determines the best placement.
```

**Review by imatwawana — COMMENTED** [review]
Thanks for updating! Made some suggestions.

File: docs/sources/panels-visualizations/visualizations/bar-gauge/index.md (hunk 1 above)
Comment (nmarrs) [reply docs/sources/panels-visualizations/visualizations/bar-gauge/index.md:78]: ah 🤦‍♂️ you are totally right - not sure where my head was at when naming this haha, it is should be called `Field label placement`

File: public/app/plugins/panel/bargauge/panelcfg.cue [hunk 4]
```diff
@@ -29,7 +29,7 @@ composableKinds: PanelCfg: {
 					common.SingleStatBaseOptions
 					displayMode:    common.BarGaugeDisplayMode & (*"gradient" | _)
 					valueMode:      common.BarGaugeValueMode & (*"color" | _)
-					valuePlacement: common.BarGaugeValuePlacement & (*"auto" | _)
+					valuePlacement: common.BarGaugeFieldLabelPlacement & (*"auto" | _)
```
Comment (adela-almasan) [inline public/app/plugins/panel/bargauge/panelcfg.cue:32]: maybe for consistency we should also rename `valuePlacement`.

**Comment by adela-almasan** [comment]
Tested locally and works well! I noticed 2 things - the option is not shown for hidden `Value display` (I don't think they're related) but it's displayed for vertical orientation although it doesn't do anything. Maybe hide it on vertical orientation and disconnect it from value display? 

https://github.com/grafana/grafana/assets/88068998/44522afa-b955-4ab5-a9e3-8fa744547194

**Comment by nmarrs** [comment]
> @nmarrs should not be named "name placement" option? it does not look like it's changing the value placement

😅 not sure what I was thinking - updated the naming to be `Field label placement` instead

File: public/app/plugins/panel/bargauge/panelcfg.cue (hunk 4 above)
Comment (nmarrs) [reply public/app/plugins/panel/bargauge/panelcfg.cue]: ah whoops missed this 😬

**Comment by leeoniya** [comment]
> 😅 not sure what I was thinking - updated the naming to be `Field label placement` instead

"Field label" might be confusin also, since fields actually have labels, and this is not that 😅 . @torkelo suggested "Name placement", which is probably better? what do we do in other panels?

**Comment by nmarrs** [comment]
> > 😅 not sure what I was thinking - updated the naming to be `Field label placement` instead
> 
> "Field label" might be confusin also, since fields actually have labels, and this is not that 😅 . @torkelo suggested "Name placement", which is probably better? what do we do in other panels?

Naming is hard 🙃 I think I would be fine with any of the following:
`fieldNamePlacement`
`gaugeLabelPlacement`
and maybe just `namePlacement` (doesn't seem as descriptive 🤷‍♂️)

**Comment by nmarrs** [comment]
> Maybe hide it on vertical orientation and disconnect it from value display?

done - thanks for catching this / the suggestion @adela-almasan  :)

File: docs/sources/panels-visualizations/visualizations/bar-gauge/index.md [hunk 5]
```diff
@@ -67,6 +67,26 @@ Choose a display mode.
...
 
+### Value display
+
+Choose a value display mode.
+
+- **Value color -** Value color is determined by value.
+- **Text color -** Value color is default text color.
+- **Hidden -** Values are hidden.
+
+### Field label placement
```
Comment (torkelo) [inline docs/sources/panels-visualizations/visualizations/bar-gauge/index.md:78]: I think name is better term than field label. In the sister panel Stat we refer to it as Name , and in settings we refer to it as "Display name" , never label.

So think "Name placement" as the option name is best

**Review by imatwawana — COMMENTED** [review]
LGTM!

**Review by imatwawana — APPROVED** [review]

**Review by adela-almasan — APPROVED** [review]
✨

**Pull request merged**